src/optimal_retention.rs

```diff
@@ -206,251 +161,172 @@ pub fn simulate(
-            .filter(|(_, &condition, _)| condition)
-            .for_each(|(rating, _, forget)| {
-                *rating = if *forget {
-                    1
-                } else {
-                    review_rating_choices[review_rating_dist.sample(&mut rng)]
-                };
-            });
-
-        // Update 'cost' column based on 'need_review', 'forget' and 'ratings'
-        izip!(&mut cost, &need_review, &forget, &ratings)
-            .filter(|(_, &need_review_flag, _, _)| need_review_flag)
-            .for_each(|(cost, _, &forget_flag, &rating)| {
-                *cost = if forget_flag {
-                    review_costs[0] * loss_aversion
-                } else {
-                    review_costs[rating - 1]
-                }
-            });
+        let rating = if forget {
+            1
+        } else {
+            review_rating_choices[review_rating_dist.sample(&mut rng)]
+        };
+
+        //dbg!(&card, &rating);
 
-        // Calculate cumulative sum of 'cost'
-        let mut cum_sum = Array1::<f32>::zeros(deck_size);
-        cum_sum[0] = cost[0];
-        for i in 1..deck_size {
-            cum_sum[i] = cum_sum[i - 1] + cost[i];
+        // Update 'cost' based on 'forget' and 'rating'
+        let cost = if not_learn {
+            if forget {
+                review_costs[0] * loss_aversion
+            } else {
+                review_costs[rating - 1]
+            }
+        } else {
+            learn_costs[rating - 1]
+        };
+
+        // Wait until a day which is available
+        while day_index < learn_span && cost_per_day[day_index] + cost > max_cost_perday {
+            day_index += 1;
+        }
+        if day_index >= learn_span {
+            card_priorities.pop();
+            continue;
         }
 
-        // Create 'true_review' mask based on 'need_review' and 'cum_sum' and 'review_limit'
-        let mut review_count = 0;
-        let true_review =
-            Zip::from(&need_review)
-                .and(&cum_sum)
-                .map_collect(|&need_review_flag, &cum_cost| {
-                    if need_review_flag {
-                        review_count += 1;
-                    }
-                    need_review_flag
-                        && (cum_cost <= max_cost_perday)
-                        && (review_count <= review_limit)
-                });
-
-        let need_learn = old_due.mapv(|x| x == learn_span as f32);
-        // Update 'cost' column based on 'need_learn'
-        izip!(&mut cost, &need_learn, &init_ratings)
-            .filter(|(_, &need_learn_flag, _)| need_learn_flag)
-            .for_each(|(cost, _, &rating)| {
-                *cost = learn_costs[rating - 1];
-            });
-
-        cum_sum[0] = cost[0];
-        for i in 1..deck_size {
-            cum_sum[i] = cum_sum[i - 1] + cost[i];
+        let retrievability =
+            power_forgetting_curve(day_index as f32 - card.last_date, card.stability);
```

Comment:
The rating is sampled from the first retrievability. However, the second retrievability is the real one.

Reply:
I ran into problems with this because 

1. The rating depends on the retrievability
1. The cost depends on the rating
2. The cost determines if the card has a cost too high to fit in the day (it might change the day)
3. The day determines the retrievability

Would it be ok to just assume the card is forgotten when checking whether the cost is too much for that day or not? Something like:
```rust
while day_index < learn_span && cost_per_day[day_index] + 
   (review_costs[0] * loss_aversion) // Card hasn't been reviewed yet, we are assuming they forgot.
   > max_cost_perday {
    day_index += 1;
}
```
To avoid rating re-roll shenanigans.

Reply:
It isn't OK to assume that. For example, assuming the original interval is 365 days, a delay of one day doesn't have an significant impact. But if the original interval is 1 day, the delay will induce a large impact.

Reply:
Oops. I'm sorry I misunderstood. It's OK to use the maximum cost when check whether the cost is too much for that day.

Reply:
I think fixing it will solve this test:

thread 'optimal_retention::tests::simulator' panicked at src/optimal_retention.rs:898:9:
assertion `left == right` failed
  left: 8334.49
 right: 6521.068

In the implementation of current PR, the memorization is significantly higher than the previous on. Because it's too optimistic to sample the rating from the first retrievability.

Reply:
With the change:

left: 5832.3096
right: 6521.068

I'm assuming the memorisation has probably gone under because we're assuming the card failed which the current version doesn't do?

Reply:
At least the value is closer to the expected value than the previous one.

I guess it's due to the `forget_cost` is so large that the review is delayed too much, which decreases the memorization.